feat(pharmacyOrders): TAN-2521: Allow users to place a pharmacy order for a patient's prescriptions

rohan-bes opened the pull request:
### Changes

This is the front-end implementation, the backend models were done in a separate PR.

- Add the UI for the Pharmacy Order modal
- Add the settings key to enable/disable pharmacy orders
- Add a route in the facility-server to allow creating a pharmacy order


### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

Commit:
feat(pharmacyOrders): TAN-2521: Add 'Order from pharmacy' button on the EncounterMedications view

Commit:
feat(pharmacyOrders): TAN-2521: Add integrations.pharmacyOrder.enabled feature flag in settings

Commit:
feat(pharmacyOrders): TAN-2521: Add Pharmacy Icon

rohan-bes commented on packages/facility-server/app/routes/apiv1/encounter.js:
```diff
@@ -186,6 +186,41 @@ encounter.post(
+  asyncHandler(async (req, res) => {
+    const { db, models, params, body } = req;
+    const { id } = params;
+    req.checkPermission('write', 'Encounter');
+    req.checkPermission('read', 'Medication');
```
Chatted with Megs and these are the permissions we settled on

Commit:
Added error handling

rohan-bes commented on packages/web/app/components/Medication/PharmacyOrderModal.jsx:
```diff
@@ -0,0 +1,532 @@
+    (event, key, rowIndex) => {
+      if ([COLUMN_KEYS.QUANTITY, COLUMN_KEYS.REPEATS].includes(key)) {
+        const newMedicationData = [...prescriptions];
+        const value =
+          key === COLUMN_KEYS.QUANTITY
+            ? parseInt(event.target.value, 10) || undefined
+            : parseInt(event.target.value, 10) || undefined;
```
We're treating undefined as 0 for repeats so I think that's okay

rohan-bes replied:
Oh you're definitely right about the ternary! This was some AI code that I edited down, and didn't clock that the two statements were the same.

Regarding the why fallback to undefined. Sorry if my comment was a bit misleading. I do kinda like the existing functionality of repeats either being null or a positive number, as it only makes sense to request a non-zero number of repeats but it's an optional field.

Commit:
Changed data-test-id

rohan-bes commented on packages/facility-server/app/routes/apiv1/encounter.js:
```diff
@@ -186,6 +186,41 @@ encounter.post(
+    const encounterObject = await models.Encounter.findByPk(id);
+    if (!encounterObject) throw new NotFoundError();
+
+    const { orderingClinicianId, comments, pharmacyOrderPrescriptions } = body;
+
+    const result = await db.transaction(async () => {
+      const pharmacyOrder = await models.PharmacyOrder.create({
+        orderingClinicianId,
+        comments,
+        encounterId: id,
+      });
+
+      for (const prescription of pharmacyOrderPrescriptions) {
+        await models.PharmacyOrderPrescription.create({
+          pharmacyOrderId: pharmacyOrder.id,
+          prescriptionId: prescription.prescriptionId,
+          quantity: prescription.quantity,
+          repeats: prescription.repeats,
+        });
+      }
+
+      return pharmacyOrder;
+    });
```
I think validating that the prescription has an EncounterPrescription for this encounter isn't a requirement. It's possible that we may want to support ongoing prescriptions too.

But will switch to bulkCreate :+1:

Commit:
Switched to using bulkCreate when creating PharmacyOrderPrescriptions

NavarroEmilioLuis reviewed (changes requested):
Kinda complex UI! I dislike having to go through so many hoops. You did a good work though! I'm only requesting changes because of that ternary operator that doesn't seem like it's doing anything? Everything else is just optional.

Commit:
PR fixups
- Removed unnecessary ternary
- Broke PharmacyOrderMedicationTable into its own component

rohan-bes commented on packages/facility-server/app/routes/apiv1/encounter.js:
```diff
@@ -186,6 +186,41 @@ encounter.post(
 
 encounter.post('/:id/createPatientLetter', createPatientLetter('Encounter', 'encounterId'));
 
+encounter.post(
```
Oh yep sure thing will add at least one

Commit:
Added unit test for POST encounter/id/pharmacyOrder route

Commit:
Merge branch 'main' into feat/tan-2521-pharmacy-order-ui

Commit:
Fixed import name after merge main

NavarroEmilioLuis commented on packages/web/app/components/Medication/PharmacyOrderModal.jsx:
```diff
@@ -0,0 +1,532 @@
+      };
+
+      await api.post(`encounter/${encounter.id}/pharmacyOrder`, orderData);
+      setShowSuccess(true);
```
Yeah I mean if we need to re-fetch the underlying table or something else? Unsure if needed

rohan-bes replied:
Ah I see, nah for now we don't display anything about the pharmacyOrders, so there's not even a `GET` endpoint for them. 

But good call cos soon we'll add a column to the Medication Table that's based on if a pharmacyOrder exists, so at that point we'll need to refetch that table when we create one here.

NavarroEmilioLuis reviewed (approved):
Happy with the changes! Sorry for not coming back earlier - I left another message about the re-fetch but it might not be needed anyway

NavarroEmilioLuis commented on packages/settings/src/schema/global.ts:
```diff
@@ -1440,9 +1450,15 @@ export const globalSettings = {
+                  ADMINISTRATION_FREQUENCIES.AS_DIRECTED,
+                ].includes(frequency),
+            ),
+          ),
```
I think it's fine! Unless we are using this filter somewhere else. I was just mainly commenting because it looked weird (but I have no context over this)

Commit:
Fixed wording on messages

Commit:
Message fixups
